### PR TITLE
add Themis contracts

### PR DESCRIPTION
**Pull request by 1szheng**
Base code for Themis contracts, for follow up add-ons

Most functionalities, other than datetime related, tested in REMIX

### CODE REVIEW

File: contracts/Campaign.sol [hunk 1]
```diff
@@ -0,0 +1,123 @@
+pragma solidity ^0.5.17;
+
+import "./IAM.sol";
+import "./CampaignFactory.sol";
+
+contract Campaign {
+    address manager;        // the factory
+    address owner;
+    uint256 lockoutEndDatetime;
```
Comment (TheSpaceCuber) [inline contracts/Campaign.sol]: Name seems a bit long, what about endDate / endDatetime etc?

File: contracts/Campaign.sol [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        address payable beneficiary = address(uint160(owner));
+        uint256 toBeDonated =  totalDonated - commission;
+
+        platform.transfer(commission);
+        beneficiary.transfer(toBeDonated);
+        emit hasWithdrawn(beneficiary, toBeDonated);
+    }
+
+    // pseudo-code for follow up    
+    function refund(address campaignAddr) public managerOnly distrustOnly {
```
Comment (TheSpaceCuber) [inline contracts/Campaign.sol]: Thanks for this placeholder function!

I'm thinking maybe the modifier distrustOnly isn't necessary as some cases the organization might want to trigger a refund even if they have not been marked as distrust.

File: contracts/Campaign.sol [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        require(isDistrustedOwner() == true);
+        _;
+    }
+
+    modifier managerOnly() {
+        require(manager == msg.sender);
+        _;
+    }
+
+    modifier withinLockoutOnly() {
```
Comment (TheSpaceCuber) [inline contracts/Campaign.sol]: Possibly confusing name if Prof sees this, what about something like
1. ongoingCampaignOnly
2. validCampaignOnly

Just some suggestions

File: contracts/CampaignFactory.sol [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    
+    function isVerified(address organisation) public view returns (bool) {
+        return IAMContract.isVerified(organisation);
+    }
+    
+    function isLocked(address organisation) public view returns (bool) {
+        return IAMContract.isLocked(organisation);
+    }
+
+    function isDistrust (address organisation) public view returns (bool) {
```
Comment (TheSpaceCuber) [inline contracts/CampaignFactory.sol]: Small issue: isDistrust (address organisation) has a white space

File: contracts/CampaignFactory.sol [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    function withdraw() public ownerOnly {
+        require(address(this).balance > 0);
+        address payable beneficiary = address(uint160(owner));
+        uint256 amount = address(this).balance;
+
+        beneficiary.transfer(amount);
+        emit commissionWithdrawn(amount);
+    }
+    
+    function refundCampaigns(address organisation) public ownerOnly {
```
Comment (TheSpaceCuber) [inline contracts/CampaignFactory.sol]: Suggestion: refundAllCampaigns may be more in line, if i understood this correctly.

File: contracts/IAM.sol [hunk 6]
```diff
@@ -0,0 +1,79 @@
+pragma solidity ^0.5.17;
+
+import "./CampaignFactory.sol";
+
+contract IAM {
+    address owner;
+    // CampaignFactory factoryContract;
```
Comment (TheSpaceCuber) [inline contracts/IAM.sol]: Why is this commented?

File: contracts/IAM.sol [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    function isDistrust(address organisation) public view returns (bool) {
+        return (orgStatus[organisation] == status.DISTRUST);
+    }
+
+    // function setFactory(CampaignFactory factoryAddress) public ownerOnly {
+    //     require(address(factoryAddress) != address(0));
+    //     factoryContract = factoryAddress;
+    // }
```
Comment (TheSpaceCuber) [inline contracts/IAM.sol]: This is for transferring ownership of the Factory is it?

**Review by TheSpaceCuber — CHANGES_REQUESTED** [review]
LGTM overall, very comprehensive, thanks for the fast updates! Just a couple of questions/comments, feel free to close them accordingly when resolved, otherwise I will close them afterwards. Will review again.

File: contracts/Campaign.sol (hunk 2 above)
Comment (1szheng) [reply contracts/Campaign.sol]: For this function, the idea was that it is only meant to be called for the full reversal of donations. Specifically in the case whereby of some kind of investigation such as fraud.

File: contracts/IAM.sol (hunk 7 above)
Comment (1szheng) [reply contracts/IAM.sol]: This was originally meant for potential callbacks, removed for clarity.

File: contracts/Campaign.sol (hunk 1 above)
Comment (1szheng) [reply contracts/Campaign.sol]: Thank you for the suggestions, updated to endDatetime.

File: contracts/Campaign.sol (hunk 3 above)
Comment (1szheng) [reply contracts/Campaign.sol]: Updated to ongoingCampaignOnly

File: contracts/CampaignFactory.sol (hunk 5 above)
Comment (1szheng) [reply contracts/CampaignFactory.sol]: Yes, the function is meant to for all campaigns of an organisation, updated as suggested.

File: contracts/IAM.sol [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    
+    function setLocked(address organisation) public ownerOnly {
+        orgStatus[organisation] = status.LOCK;
+        emit orgLocked(organisation);
+    }
+    
+    function setDistrust(address organisation) public ownerOnly {
+        orgStatus[organisation] = status.DISTRUST;
+        emit orgDistrust(organisation);
+    }
```
Comment (alanlowzies) [inline contracts/IAM.sol]: Comprehensive getters and setters 👍🏻

File: contracts/Campaign.sol (hunk 2 above)
Comment (alanlowzies) [reply contracts/Campaign.sol]: Yeah I think for now it will be safer for the 'mass refund' function to be only used for fraud instances.

If we want to go with one of our KIV plan (or is it to be implemented?) to allow donors to withdraw their donations while the campaign is still active, perhaps we can have another function for that?

File: contracts/CampaignFactory.sol [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    address owner;
+    IAM IAMContract;
+    uint8 MAX_CHARITIES = 5;
+    uint16 HoursInYear = 8760;
+    uint16 SecsInHour = 3600;
+
+    mapping(address => address[]) orgCampaigns;
+
+    event mountCampaign(address org, address campaign, uint256 durationHrs);
+    event campaignEnded(address org, address campaign);
```
Comment (alanlowzies) [inline contracts/CampaignFactory.sol]: For consistency sake, perhaps `org` can be expanded to `organisation`?

File: contracts/CampaignFactory.sol [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+            if (orgCampaigns[organisation][i+1] != address(campaign)) {
+                orgCampaigns[organisation][i] = orgCampaigns[organisation][i+1];
+            }
+        }
+        orgCampaigns[organisation].pop();
+        emit campaignEnded(organisation, address(campaign));
+    }
+
+    // withdraw commission
+    function withdraw() public ownerOnly {
```
Comment (alanlowzies) [inline contracts/CampaignFactory.sol]: I'm assuming this function is to transfer all commissions received from past campaigns to date to the owner of `CampaignFactory`.

For better clarity, perhaps the function name can be more concise? For example, `sendCommissionsToOwner`. A bit wordy, but I think the clarity would help.

File: contracts/CampaignFactory.sol [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+        require(address(this).balance > 0);
+        address payable beneficiary = address(uint160(owner));
```
Comment (alanlowzies) [inline contracts/CampaignFactory.sol]: Perhaps the variable name can be changed to something like `ownerAddr`? The name `beneficiary` can be confusing since we are not transferring the commission to a beneficiary but rather the owner of `CampaignFactory` (unless I misunderstood the purpose of the function).

File: contracts/CampaignFactory.sol [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        require(durationHrs <= HoursInYear);
+        require(orgCampaigns[msg.sender].length < MAX_CHARITIES);
+        
+        uint256 durationSecs = hoursToSeconds(durationHrs);
+        Campaign c = new Campaign(durationSecs, msg.sender, IAMContract);
+        orgCampaigns[msg.sender].push(address(c));
+        
+        emit mountCampaign(msg.sender, address(c), durationHrs);
+        return c;
+    }
```
Comment (alanlowzies) [inline contracts/CampaignFactory.sol]: Interesting, I didn't know Solidity supported overloading of functions. TIL

File: contracts/CampaignFactory.sol [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+import "./IAM.sol";
+
+contract CampaignFactory {
+    address owner;
+    IAM IAMContract;
+    uint8 MAX_CHARITIES = 5;
+    uint16 HoursInYear = 8760;
+    uint16 SecsInHour = 3600;
+
+    mapping(address => address[]) orgCampaigns;
```
Comment (alanlowzies) [inline contracts/CampaignFactory.sol:13]: Am I right to assume that this list will only contain organisations/beneficiaries with active campaigns? That is, if an organisation no longer has any active campaign, it will be removed from the `orgCampaigns` mapping?

File: contracts/CampaignFactory.sol [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+        for (int i = int(len); i >= 0; i--) {
+            Campaign c = Campaign(orgCampaigns[organisation][uint256(i)]);
+            c.refund(address(c)); // placeholder
+            deleteCampaign(organisation, address(c));
+        }
+        deleteOrganisation(organisation);
+        emit refundComplete(organisation);
+    }
+
+    function deleteCampaign(address organisation, address campaignAddr) private {
```
Comment (alanlowzies) [inline contracts/CampaignFactory.sol]: For better clarity, perhaps the function name can be `deleteCampaignFromMapping`?

File: contracts/CampaignFactory.sol [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+        orgCampaigns[organisation].pop();
+        emit campaignDeleted(organisation, campaignAddr);
+    }
+
+    function deleteOrganisation(address organisation) private {
```
Comment (alanlowzies) [inline contracts/CampaignFactory.sol]: For better clarity, perhaps the function name can be `deleteOrgFromMapping`?

File: contracts/Campaign.sol [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+
+    function isPastLockout() public view returns (bool) {
+        return block.timestamp >= endDatetime;
+    }
+
+    function getCharityOrganisation() public view returns (address) {
+        return owner;
+    }
+
+    function getendDatetime() public view returns (uint256) {
```
Comment (alanlowzies) [inline contracts/Campaign.sol]: I think you forgot to capitalise the 'e' in `end`. Sorry OCD 😅

File: contracts/Campaign.sol [hunk 17]
```diff
@@ -0,0 +1,123 @@
+pragma solidity ^0.5.17;
+
+import "./IAM.sol";
+import "./CampaignFactory.sol";
+
+contract Campaign {
+    address manager;        // the factory
```
Comment (alanlowzies) [inline contracts/Campaign.sol]: If the address will always be `CampaignFactory`'s address (i.e. `CampaignFactory` contract will always be the one creating new `Campaign` contracts), perhaps the variable name can be changed to `campaignFactory`. I feel that it will be more understandable.

File: contracts/Campaign.sol [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+        totalDonated += msg.value;
+        emit donationMade(msg.sender, msg.value);
+    }
+
+    function withdraw() public ownerOnly verifiedOnly pastLockoutOnly {
+        uint256 commission = (totalDonated * commissionBP) / basispoints;
+        
+        address payable platform = address(uint160(manager));
+        address payable beneficiary = address(uint160(owner));
+        uint256 toBeDonated =  totalDonated - commission;
```
Comment (alanlowzies) [inline contracts/Campaign.sol]: Perhaps a better variable name could be `netDonationAmt`?

File: contracts/Campaign.sol [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+    function donate() public payable verifiedOnly ongoingCampaignOnly {
+        require(msg.value > 0);
+        totalDonated += msg.value;
+        emit donationMade(msg.sender, msg.value);
+    }
+
+    function withdraw() public ownerOnly verifiedOnly pastLockoutOnly {
+        uint256 commission = (totalDonated * commissionBP) / basispoints;
+        
+        address payable platform = address(uint160(manager));
```
Comment (alanlowzies) [inline contracts/Campaign.sol]: For clarity, perhaps the `platform` variable can be renamed to `campaignFactory`? 

Although the money that will be transferred is for the upkeep of the platform, since we will be transferring it to the `campaignFactory` contract, it may be clearer to name the variable that.

**Review by alanlowzies — CHANGES_REQUESTED** [review]
The logic for most of the code in the three contracts looks good! 👍🏻 Most of my comments are suggestions to rename certain function and variable names to aid in their clarity and consistency. I do have one or two questions about some parts that I am unsure of and I would appreciate your help in understanding them.

A general suggestion I have for your consideration is to maybe include error messages with `require` function calls so that if a transaction reverts a reason will be shown.

All in all, thanks for writing these contracts, they're pretty solid (no pun intended 🌚).

(My apologies if some of my comments are unclear, I did this late at night haha. Please let me know if you need some clarifications on my comments.)

File: contracts/Campaign.sol (hunk 2 above)
Comment (1szheng) [reply contracts/Campaign.sol]: I believe that we currently not looking to implement that version of withdrawal as it could open windows for tax fraud, but is a KIV for future considerations.

File: contracts/CampaignFactory.sol (hunk 9 above)
Comment (1szheng) [reply contracts/CampaignFactory.sol]: Updated variable names as suggested

File: contracts/CampaignFactory.sol (hunk 10 above)
Comment (1szheng) [reply contracts/CampaignFactory.sol]: Updated to `withdrawCommissions` to be a little more brief.

File: contracts/CampaignFactory.sol (hunk 11 above)
Comment (1szheng) [reply contracts/CampaignFactory.sol]: Updated variable name as suggested

File: contracts/CampaignFactory.sol (hunk 13 above)
Comment (1szheng) [reply contracts/CampaignFactory.sol:13]: Yes, that is indeed the case - the removal is performed by `closeCampaign()` called from Campaign.sol's `withdraw()`

File: contracts/CampaignFactory.sol (hunk 14 above)
Comment (1szheng) [reply contracts/CampaignFactory.sol]: Updated function name as suggested

File: contracts/CampaignFactory.sol (hunk 15 above)
Comment (1szheng) [reply contracts/CampaignFactory.sol]: Updated function name as suggested

File: contracts/Campaign.sol (hunk 16 above)
Comment (1szheng) [reply contracts/Campaign.sol]: Good catch, typo from mass replacement

File: contracts/Campaign.sol (hunk 17 above)
Comment (1szheng) [reply contracts/Campaign.sol]: Good point, was thinking of it more from a business-esque view, updated to `campaignFactory`

File: contracts/Campaign.sol (hunk 18 above)
Comment (1szheng) [reply contracts/Campaign.sol]: Updated variable name as suggested

File: contracts/Campaign.sol (hunk 19 above)
Comment (1szheng) [reply contracts/Campaign.sol]: Variable name updated to `campgnFactory` to avoid clash with updated `manager` name

**Review by TheSpaceCuber — APPROVED** [review]
LGTM

File: contracts/CampaignFactory.sol (hunk 13 above)
Comment (alanlowzies) [reply contracts/CampaignFactory.sol:13]: I see. Alright, thanks for the clarification!

**Review by alanlowzies — APPROVED** [review]
Thanks for the prompt changes. I'm okay with the contracts.
